pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +238,267 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+		sectionCycleEnd, _ = nextCycleEnd(
+			state.CommitRecTimestamp,
+			b.cfg.ConsumeInterval,
+			b.cfg.ConsumeIntervalBuffer,
+		)
+		level.Info(b.logger).Log("msg", "partition is lagging behind the cycle", "partition", partition, "lag", state.Lag, "section_cycle_end", sectionCycleEnd, "cycle_end", cycleEnd, "commit_rec_ts", state.CommitRecTimestamp)
+	}
+	for !sectionCycleEnd.After(cycleEnd) {
+		partitionLogger := log.With(b.logger, "partition", partition, "lag", state.Lag, "section_cycle_end", sectionCycleEnd)
+		state, err = b.consumePartitionCycle(ctx, partitionLogger, builder, partition, state, sectionCycleEnd)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", partition, err)
+		}
+		sectionCycleEnd = sectionCycleEnd.Add(b.cfg.ConsumeInterval)
+	}
+
+	return nil
+}
+
+// consumePartition consumes records from the given partition until the cycleEnd timestamp.
+// If the partition is lagging behind, the caller needs to take care of calling consumePartition in sections.
+func (b *BlockBuilder) consumePartitionCycle(ctx context.Context, logger log.Logger, builder *TSDBBuilder, partition int32, state partitionState, cycleEnd time.Time) (_ partitionState, retErr error) {
+	blockEnd := cycleEnd.Truncate(b.cfg.ConsumeInterval)
+
+	var numBlocks int
+	defer func(t time.Time, startState partitionState) {
+		dur := time.Since(t)
+
+		// Don't propagate any errors derived from the canceled context.
+		if errors.Is(retErr, context.Canceled) {
+			retErr = nil
+		}
+
+		if retErr != nil {
+			level.Error(logger).Log("msg", "partition consumption failed", "duration", dur, "curr_lag", state.Lag, "err", retErr)
+			return
+		}
+
+		b.blockBuilderMetrics.processPartitionDuration.WithLabelValues(fmt.Sprintf("%d", partition)).Observe(dur.Seconds())
+		level.Info(logger).Log("msg", "done consuming", "duration", dur, "curr_lag", state.Lag,
+			"last_block_end", startState.LastBlockEnd, "curr_block_end", blockEnd,
+			"last_seen_offset", startState.LastSeenOffset, "curr_seen_offset", state.LastSeenOffset,
+			"num_blocks", numBlocks)
+	}(time.Now(), state)
+
+	// We always rewind the partition's offset to the commit offset by reassigning the partition to the client (this triggers partition assignment).
+	// This is so the cycle started exactly at the commit offset, and not at what was (potentially over-) consumed previously.
+	b.kafkaClient.AddConsumePartitions(map[string]map[int32]kgo.Offset{
+		b.cfg.Kafka.Topic: {
+			partition: kgo.NewOffset().At(state.Commit.At),
+		},
+	})
+	defer b.kafkaClient.RemoveConsumePartitions(map[string][]int32{b.cfg.Kafka.Topic: {partition}})
+
+	level.Info(logger).Log("msg", "start consuming", "offset", state.Commit.At)
+
+	var (
+		firstRec  *kgo.Record
+		lastRec   *kgo.Record
+		commitRec *kgo.Record
+	)
+
+consumerLoop:
+	for remaining := state.Lag; remaining > 0; {
```

Comment:
Generally speaking, the `remaining > 0` is always a red flag to me, because it's hard to reason about possible edge cases where `remaining > 0` but we actually consumed the partition till the end (i.e. if `state.Lag` was not computed correctly). I'm thinking in particular to edge cases related to inactive partitions, where no new records are produced, and so this `for` loop may potentially never end.

To prevent for such starvation in case of issues we're not considering right now, what if we pass to `PollFetches()` a context with a long timeout (e.g. few minutes) and we break the consumption (like if `remaining` has reached 0) if such timeout expires?

Reply:
> what if we pass to `PollFetches()` a context with a long timeout (e.g. few minutes) and we break the consumption (like if remaining has reached 0) if such timeout expires?

This was exactly what I did in the earlier versions. I bumped into the problem that `PollFetches` can randomly return an empty set of fetches if the broker decides so. If we get rid of the `remaining` guard, we cannot tell if we should bail out from the polling loop (because we consumed the whole backlog), or we must retry the polling.

Reply:
> I bumped into the problem that PollFetches can randomly return an empty set of fetches if the broker decides so.

Interesting. Add a code comment to explain it, please.